File: backend/Bookshelf.Application/Common/PasswordPolicy.cs
using System.Text.RegularExpressions;
using Bookshelf.Application.Common.Exceptions;

namespace Bookshelf.Application.Common;

/// <summary>
/// Shared password strength rule for registration and password changes — kept in one place
/// so the two call sites (and the frontend hint text) can never drift out of sync.
/// </summary>
public static partial class PasswordPolicy
{
    public const string RequirementsMessage =
        "Password must be at least 8 characters and include an uppercase letter, a lowercase letter, a number, and a special character.";

    public static void Validate(string password)
    {
        if (!Regex().IsMatch(password))
        {
            throw new ValidationException(RequirementsMessage);
        }
    }

    [GeneratedRegex(@"^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[^a-zA-Z0-9]).{8,}$")]
    private static partial Regex Regex();
}

File: frontend/src/app/core/validators/password-policy.ts
/**
 * Mirrors backend/Bookshelf.Application/Common/PasswordPolicy.cs exactly — keep both in
 * sync if the rule ever changes, so the frontend hint never promises something the
 * backend doesn't actually enforce (or vice versa).
 */
export const PASSWORD_PATTERN = /^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[^a-zA-Z0-9]).{8,}$/;

export const PASSWORD_REQUIREMENTS_HINT =
  'At least 8 characters, with an uppercase letter, a lowercase letter, a number, and a special character.';
